### PR TITLE
add compute functions

### PR DESCRIPTION
**Pull request by andreaturli**
- fix Device and PaginatedCollection

### CODE REVIEW

File: packet/src/main/java/org/jclouds/packet/compute/functions/DeviceToNodeMetadata.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    private final OperatingSystemToImage operatingSystemToImage;
+    private final FacilityToLocation facilityToLocation;
+    private final Function<Device.State, NodeMetadata.Status> toPortableStatus;
+    private final GroupNamingConvention groupNamingConvention;
+
+    @Inject
+    DeviceToNodeMetadata(PlanToHardware planToHardware, OperatingSystemToImage operatingSystemToImage, FacilityToLocation facilityToLocation,
+                         Function<Device.State, NodeMetadata.Status> toPortableStatus,
+                         GroupNamingConvention.Factory groupNamingConvention) {
+        this.planToHardware = checkNotNull(planToHardware, "planToHardware cannot be null");
```
Comment (nacx) [inline packet/src/main/java/org/jclouds/packet/compute/functions/DeviceToNodeMetadata.java]: No need for null checks in the injection constructors.

File: packet/src/main/java/org/jclouds/packet/compute/functions/FacilityToLocation.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        this.justProvider = justProvider;
+    }
+
+    @Override
+    public Location apply(final Facility facility) {
+        final LocationBuilder builder = new LocationBuilder();
+        builder.id(facility.code());
+        builder.description(facility.name());
+        builder.parent(getOnlyElement(justProvider.get()));
+        builder.scope(LocationScope.REGION);
```
Comment (nacx) [inline packet/src/main/java/org/jclouds/packet/compute/functions/FacilityToLocation.java]: Can the ISO-3166 code be inferred in a consistent way, for example from the `state` field?

File: packet/src/main/java/org/jclouds/packet/compute/functions/PlanToHardware.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                .name(plan.name())
+                .hypervisor("none")
+                .processors(getProcessors(plan))
+                .ram(getMemory(plan))
+                .volumes(getVolumes(plan));
+        return builder.build();
+    }
+
+    private Integer getMemory(Plan plan) {
+        if (plan.specs() == null || plan.specs().drives() == null) return -1;
```
Comment (nacx) [inline packet/src/main/java/org/jclouds/packet/compute/functions/PlanToHardware.java]: Better return 0 ?

File: packet/src/main/java/org/jclouds/packet/compute/functions/FacilityToLocation.java (hunk 2 above)
Comment (andreaturli) [inline packet/src/main/java/org/jclouds/packet/compute/functions/FacilityToLocation.java]: it'd be a bit tricky as the API returns https://github.com/jclouds/jclouds-labs/blob/master/packet/src/test/resources/facilities.json and it is not straightforward to get the correct ISO-3166 from the `name` and/or `code` I think

Comment (nacx) [inline packet/src/main/java/org/jclouds/packet/compute/functions/FacilityToLocation.java]: I was referring to the `state` field (according to their docs). Are there new fields that have been added recently on their side?
https://www.packet.net/developers/api/devices/#devices-locations

Comment (andreaturli) [inline packet/src/main/java/org/jclouds/packet/compute/functions/FacilityToLocation.java]: I think the docs is not in sync: there is no `state` in the live json object :(

Comment (nacx) [inline packet/src/main/java/org/jclouds/packet/compute/functions/FacilityToLocation.java]: Let's leave it as-is then. Thanks for clarifying!

**Comment by andreaturli** [comment]
squashing and merging

**Comment by andreaturli** [comment]
merged at [master](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/a91ff3bd)

**Pull request closed without merging**